File: web/lib/schemas.ts
/**
 * Zod input schemas: one per FastAPI endpoint.
 *
 * Every field name is a **real column from the real dataset**, mirroring
 * `src/serving/schemas.py` exactly. A mismatch here is a 422 from FastAPI, so
 * these are the contract, not a convenience layer.
 *
 * They are deliberate subsets: IEEE-CIS has 394 columns and a 394-input form is
 * not a demo. Unsupplied columns arrive at the model as NaN, which LightGBM
 * handles natively as "unknown"; see `src/serving/preprocessing.py`.
 *
 * Defaults are seeded demo values, never real customer data.
 */
import { z } from "zod";

// ─── Payment fraud (IEEE-CIS) ──────────────────────────────────────────────

export const FraudInputSchema = z.object({
  // Seconds from the dataset's reference point, NOT a Unix timestamp.
  TransactionDT: z.number().int().min(0),
  TransactionAmt: z.number().positive(),
  ProductCD: z.enum(["W", "C", "R", "H", "S"]),
  card1: z.number(),
  card2: z.number(),
  card3: z.number(),
  card4: z.enum(["visa", "mastercard", "discover", "american express"]),
  card5: z.number(),
  card6: z.enum(["debit", "credit"]),
  addr1: z.number(),
  addr2: z.number(),
  dist1: z.number(),
  P_emaildomain: z.string().min(1),
  C1: z.number().min(0),
  C13: z.number().min(0),
  C14: z.number().min(0),
  D1: z.number(),
  D15: z.number(),
});
export type FraudInput = z.infer<typeof FraudInputSchema>;

export const FRAUD_DEFAULTS: FraudInput = {
  TransactionDT: 8_640_000,
  TransactionAmt: 149.99,
  ProductCD: "W",
  card1: 13926,
  card2: 361,
  card3: 150,
  card4: "visa",
  card5: 226,
  card6: "debit",
  addr1: 315,
  addr2: 87,
  dist1: 19,
  P_emaildomain: "gmail.com",
  C1: 1,
  C13: 1,
  C14: 1,
  D1: 14,
  D15: 0,
};

// ─── Consumer credit risk (LendingClub) ────────────────────────────────────
// Every field is available at origination. No post-origination field appears
// here. That is the leak configs/credit_risk.yaml's denylist exists to prevent.

export const CreditRiskInputSchema = z.object({
  loan_amnt: z.number().positive(),
  funded_amnt: z.number().positive(),
  term: z.enum([" 36 months", " 60 months"]),
  int_rate: z.number().min(0).max(50),
  installment: z.number().positive(),
  grade: z.enum(["A", "B", "C", "D", "E", "F", "G"]),
  sub_grade: z.string().regex(/^[A-G][1-5]$/, "Format: a grade letter then 1-5, e.g. C1"),
  emp_length: z.string().min(1),
  home_ownership: z.enum(["RENT", "OWN", "MORTGAGE"]),
  annual_inc: z.number().positive(),
  verification_status: z.enum(["Verified", "Source Verified", "Not Verified"]),
  purpose: z.string().min(1),
  addr_state: z.string().length(2),
  dti: z.number().min(0),
  delinq_2yrs: z.number().min(0),
  fico_range_low: z.number().min(300).max(850),
  fico_range_high: z.number().min(300).max(850),
  inq_last_6mths: z.number().min(0),
  open_acc: z.number().min(0),
  pub_rec: z.number().min(0),
  revol_bal: z.number().min(0),
  revol_util: z.number().min(0),
  total_acc: z.number().min(0),
  application_type: z.enum(["Individual", "Joint App"]),
  mort_acc: z.number().min(0),
  pub_rec_bankruptcies: z.number().min(0),
});
export type CreditRiskInput = z.infer<typeof CreditRiskInputSchema>;

export const CREDIT_RISK_DEFAULTS: CreditRiskInput = {
  loan_amnt: 15000,
  funded_amnt: 15000,
  term: " 36 months",
  int_rate: 13.56,
  installment: 509.66,
  grade: "C",
  sub_grade: "C1",
  emp_length: "5 years",
  home_ownership: "MORTGAGE",
  annual_inc: 72000,
  verification_status: "Source Verified",
  purpose: "debt_consolidation",
  addr_state: "CA",
  dti: 18.24,
  delinq_2yrs: 0,
  fico_range_low: 695,
  fico_range_high: 699,
  inq_last_6mths: 1,
  open_acc: 11,
  pub_rec: 0,
  revol_bal: 14300,
  revol_util: 52.4,
  total_acc: 24,
  application_type: "Individual",
  mort_acc: 1,
  pub_rec_bankruptcies: 0,
};

// ─── Card attrition ────────────────────────────────────────────────────────
// The two Naive_Bayes_Classifier_* columns from the published CSV are absent on
// purpose: they are the target laundered through a classifier and are denylisted
// in configs/churn.yaml.

export const ChurnInputSchema = z.object({
  Customer_Age: z.number().min(18).max(120),
  Gender: z.enum(["M", "F"]),
  Dependent_count: z.number().min(0),
  Education_Level: z.enum([
    "Uneducated",
    "High School",
    "College",
    "Graduate",
    "Post-Graduate",
    "Doctorate",
    "Unknown",
  ]),
  Marital_Status: z.enum(["Married", "Single", "Divorced", "Unknown"]),
  Income_Category: z.enum([
    "Less than $40K",
    "$40K - $60K",
    "$60K - $80K",
    "$80K - $120K",
    "$120K +",
    "Unknown",
  ]),
  Card_Category: z.enum(["Blue", "Silver", "Gold", "Platinum"]),
  Months_on_book: z.number().min(0),
  Total_Relationship_Count: z.number().min(0),
  Months_Inactive_12_mon: z.number().min(0).max(12),
  Contacts_Count_12_mon: z.number().min(0),
  Credit_Limit: z.number().positive(),
  Total_Revolving_Bal: z.number().min(0),
  Avg_Open_To_Buy: z.number().min(0),
  Total_Amt_Chng_Q4_Q1: z.number().min(0),
  Total_Trans_Amt: z.number().min(0),
  Total_Trans_Ct: z.number().min(0),
  Total_Ct_Chng_Q4_Q1: z.number().min(0),
  Avg_Utilization_Ratio: z.number().min(0).max(1),
});
export type ChurnInput = z.infer<typeof ChurnInputSchema>;

export const CHURN_DEFAULTS: ChurnInput = {
  Customer_Age: 45,
  Gender: "M",
  Dependent_count: 2,
  Education_Level: "Graduate",
  Marital_Status: "Married",
  Income_Category: "$60K - $80K",
  Card_Category: "Blue",
  Months_on_book: 36,
  Total_Relationship_Count: 4,
  Months_Inactive_12_mon: 2,
  Contacts_Count_12_mon: 3,
  Credit_Limit: 12000,
  Total_Revolving_Bal: 1200,
  Avg_Open_To_Buy: 10800,
  Total_Amt_Chng_Q4_Q1: 0.75,
  Total_Trans_Amt: 4400,
  Total_Trans_Ct: 67,
  Total_Ct_Chng_Q4_Q1: 0.68,
  Avg_Utilization_Ratio: 0.1,
};

File: src/serving/schemas.py
"""Request schemas for the three prediction endpoints.

Every field name is a **real column from the real dataset**, not a friendlier
invention. That matters: a reviewer reading ``/predict/fraud`` should be able to
open the IEEE-CIS data description and find ``card1`` and ``ProductCD`` there. It
also means the serving path can hand the payload straight to the same feature code
training used, with no translation layer to drift out of sync.

These are deliberate *subsets*. IEEE-CIS has 394 columns; a form with 394 inputs is
not a demo. Unsupplied columns arrive at the model as NaN, which LightGBM handles
natively as "unknown"; see ``src/serving/preprocessing.py``.

Defaults are seeded demo values, never real customer data.
"""

from __future__ import annotations

from pydantic import BaseModel, Field


class TransactionRequest(BaseModel):
    """A payment, in IEEE-CIS column names.

    ``TransactionDT`` is the dataset's seconds-from-reference offset, not a Unix
    timestamp. The default corresponds to roughly day 100 of the dataset window.
    """

    TransactionDT: int = Field(8_640_000, description="Seconds from the dataset reference point")
    TransactionAmt: float = Field(149.99, gt=0, description="Transaction amount")
    ProductCD: str = Field("W", description="Product code: W, C, R, H or S")
    card1: float = Field(13926, description="Anonymised card identifier")
    card2: float | None = Field(361.0, description="Anonymised card attribute")
    card3: float = Field(150.0, description="Anonymised card attribute")
    card4: str = Field("visa", description="Card network")
    card5: float = Field(226.0, description="Anonymised card attribute")
    card6: str = Field("debit", description="debit or credit")
    addr1: float = Field(315.0, description="Anonymised billing region")
    addr2: float = Field(87.0, description="Anonymised billing country")
    dist1: float | None = Field(19.0, description="Anonymised distance")
    P_emaildomain: str = Field("gmail.com", description="Purchaser email domain")
    R_emaildomain: str | None = Field(None, description="Recipient email domain")
    C1: float = Field(1.0, description="Counting feature")
    C13: float = Field(1.0, description="Counting feature")
    C14: float = Field(1.0, description="Counting feature")
    D1: float | None = Field(14.0, description="Timedelta feature (days)")
    D15: float | None = Field(0.0, description="Timedelta feature (days)")


class LoanApplicationRequest(BaseModel):
    """A loan application, in LendingClub column names.

    Every field is available at origination. No post-origination field appears
    here or anywhere in the serving path. That is the leak
    ``configs/credit_risk.yaml``'s denylist exists to prevent.
    """

    loan_amnt: float = Field(15000, gt=0, description="Requested amount")
    funded_amnt: float = Field(15000, gt=0, description="Funded amount")
    term: str = Field(" 36 months", description="' 36 months' or ' 60 months'")
    int_rate: float = Field(13.56, description="Interest rate, percent")
    installment: float = Field(509.66, gt=0, description="Monthly payment")
    grade: str = Field("C", description="LendingClub grade A-G")
    sub_grade: str = Field("C1", description="LendingClub sub-grade, e.g. C1")
    emp_length: str = Field("5 years", description="Employment length as reported")
    home_ownership: str = Field("MORTGAGE", description="RENT, OWN or MORTGAGE")
    annual_inc: float = Field(72000, gt=0, description="Self-reported annual income")
    verification_status: str = Field("Source Verified", description="Income verification")
    purpose: str = Field("debt_consolidation", description="Stated loan purpose")
    addr_state: str = Field("CA", description="Two-letter state")
    dti: float = Field(18.24, description="Debt-to-income ratio")
    delinq_2yrs: float = Field(0, ge=0, description="Delinquencies in the last 2 years")
    fico_range_low: float = Field(695, description="Lower bound of the FICO band")
    fico_range_high: float = Field(699, description="Upper bound of the FICO band")
    inq_last_6mths: float = Field(1, ge=0, description="Credit inquiries, last 6 months")
    open_acc: float = Field(11, ge=0, description="Open credit lines")
    pub_rec: float = Field(0, ge=0, description="Derogatory public records")
    revol_bal: float = Field(14300, ge=0, description="Revolving balance")
    revol_util: float = Field(52.4, description="Revolving utilisation, percent")
    total_acc: float = Field(24, ge=0, description="Total credit lines")
    application_type: str = Field("Individual", description="Individual or Joint App")
    mort_acc: float = Field(1, ge=0, description="Mortgage accounts")
    pub_rec_bankruptcies: float = Field(0, ge=0, description="Public-record bankruptcies")


class CardholderRequest(BaseModel):
    """A credit-card customer, in the attrition dataset's column names.

    The two ``Naive_Bayes_Classifier_*`` columns from the published CSV are
    absent on purpose: they are the target laundered through a classifier, and
    they are denylisted in ``configs/churn.yaml``.
    """

    Customer_Age: float = Field(45, ge=18, description="Age in years")
    Gender: str = Field("M", description="M or F")
    Dependent_count: float = Field(2, ge=0, description="Number of dependents")
    Education_Level: str = Field("Graduate", description="Education level as reported")
    Marital_Status: str = Field("Married", description="Marital status as reported")
    Income_Category: str = Field("$60K - $80K", description="Income band as reported")
    Card_Category: str = Field("Blue", description="Blue, Silver, Gold or Platinum")
    Months_on_book: float = Field(36, ge=0, description="Relationship length in months")
    Total_Relationship_Count: float = Field(4, ge=0, description="Products held")
    Months_Inactive_12_mon: float = Field(2, ge=0, description="Inactive months, last 12")
    Contacts_Count_12_mon: float = Field(3, ge=0, description="Service contacts, last 12")
    Credit_Limit: float = Field(12000, gt=0, description="Credit limit")
    Total_Revolving_Bal: float = Field(1200, ge=0, description="Revolving balance")
    Avg_Open_To_Buy: float = Field(10800, ge=0, description="Average available credit")
    Total_Amt_Chng_Q4_Q1: float = Field(0.75, description="Spend change Q4 vs Q1")
    Total_Trans_Amt: float = Field(4400, ge=0, description="Total transaction amount")
    Total_Trans_Ct: float = Field(67, ge=0, description="Total transaction count")
    Total_Ct_Chng_Q4_Q1: float = Field(0.68, description="Count change Q4 vs Q1")
    Avg_Utilization_Ratio: float = Field(0.1, ge=0, le=1, description="Average utilisation")


#: Route path fragment -> (problem name, request schema).
ROUTES = {
    "fraud": ("fraud", TransactionRequest),
    "credit-risk": ("credit_risk", LoanApplicationRequest),
    "churn": ("churn", CardholderRequest),
}
